src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -0,0 +1,405 @@
+  std::uint64_t uptime = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
+  time_t sec_off = uptime / BILLION;
+  ts.tv_sec -= sec_off;
+  uptime -= sec_off * BILLION;
+  ts.tv_usec -= uptime;
+
+  logger()->info("Creating Packetcapture instance");
+    setCapture(conf.getCapture());
+    setAnomimize(conf.getAnomimize());
+
+  if (conf.dumpIsSet()) {
+    setDump(conf.getDump());
+  }
+
+  setNetworkmode(conf.getNetworkmode());
+  addFilters(conf.getFilters());
+  addGlobalheader(conf.getGlobalheader());
+  CapStatus = (uint8_t) conf.getCapture();
+
+  auto t_filters_in = get_array_table<filters_table>("filters_map", 0, ProgramType::INGRESS);
+  auto t_filters_out = get_array_table<filters_table>("filters_map", 0, ProgramType::EGRESS);
+  filters_table ft_init;
+  ft_init.dst_port_flag = ft_init.src_port_flag = ft_init.network_filter_src_flag = ft_init.network_filter_dst_flag = ft_init.l4proto_flag = false;
+  t_filters_in.set(0x0, ft_init);
+  t_filters_out.set(0x0, ft_init);
+  writeHeader = true;
+  dt = "";
+}
+
+void Packetcapture::writeDump(const std::vector<uint8_t> &packet){
+
+  PacketJsonObject pj;
+  auto p = std::shared_ptr<Packet>(new Packet(*this, pj));
+  std::streamsize len;
+
+  if (dt == "") {
+    time_t now = time(0);
+    tm *ltm = localtime(&now);
+    std::stringstream tmstmp;
+    tmstmp << ltm->tm_mday << "-" << ltm->tm_mon << "-" << (ltm->tm_year + 1900) << "-" << ltm->tm_hour << ":" << ltm->tm_min << ":" << ltm->tm_sec;
+    dt = tmstmp.str();
+  }
+  
+  myFile.open("capture_"+ dt +".pcap", std::ios::binary | std::ios::app);
+
+  if (writeHeader == true) {
+    struct pcap_file_header *pcap_header = new struct pcap_file_header;
+    pcap_header->magic = global_header->getMagic();
+    pcap_header->version_major = global_header->getVersionMajor();
+    pcap_header->version_minor = global_header->getVersionMinor();
+    pcap_header->thiszone = global_header->getThiszone();   //timestamp are always in GMT
+    pcap_header->sigfigs = global_header->getSigfigs();
+    pcap_header->snaplen = filters->getSnaplen();
+    pcap_header->linktype = global_header->getLinktype();
+
+    myFile.write(reinterpret_cast<const char*>(pcap_header), sizeof(*pcap_header));
+    writeHeader = false;
+  }
+
+  /*
+   * Here the packet capture time offset must be added to the system boot time stored in epoch format.
+   * See line 61 of this file to see system boot time stored in epoch algorithm
+   */
+
+  struct timeval tp;
+  tp.tv_sec = ts.tv_sec;
+  tp.tv_usec = ts.tv_usec;
+  time_t sec_off = temp_offset/BILLION;     /* from nanoseconds to seconds */
+  temp_offset -= sec_off*BILLION;
+  tp.tv_usec += temp_offset/ONE_THOUSAND;   /* from nanoseconds to microseconds */
+  tp.tv_sec += sec_off;
+  sec_off = tp.tv_usec/MILLION;             /* from microseconds to seconds */
+  tp.tv_usec -= sec_off*MILLION;
+  tp.tv_sec += sec_off;
```

Comment:
A general comment to this code here. Why don't you try to use more C++ features to make the code more readable?

For instance, instead of saving the `struct timeval ts` as a member of this class, you could save the C++ `timepoint` object; then after calculating the difference between the two objects you could extract the secs and microseconds that you need to insert into the pcap header.
This would make the code more readable and portable (even if it is not really needed in our case).

Reply:
Well, I changed this part.
Thank you for your help @sebymiano